fix some issues around broadcast profiles

iamcodemaker opened the pull request:
This fixes the following issues:
- crash when deleting profiles (introduced in 10ec4cb).
- proper handling of profiles with dot in the name
- proper handling of profiles with forbidden characters in the name

Previously the code assumed the profile name was stored in the xml file but this was not the case. This has been fixed. Additionally, the code would derive the file name from the profile name which doesn't work if the two don't match (like in the previously mentioned case with a `.` in the name). Now the filename is tracked separately.

There are still some edge cases that may cause unexpected behavior if a user previously had profiles with dots or special characters in the name. Specifically, the profile name in the GUI won't match the file name and it will be possible to have several profiles with the same name  (the last one loaded will be used). This should only happen if a user previously saved a profile with a dot (`.`) in the name. Going forward, any new or edited profiles should behave as expected.

iamcodemaker committed:
use QMap::erase() to delete profiles

Back in 10ec4cb3d6c5d6a3d7727ef74a16c7c3526ed30f this was changed from
iterating over `m_profiles.values()` to iterating over `m_profiles`
directly. This causes problems because we modify the structure we are
iterating over. The fix here is using `erase()` to do the deleting.
Alternatively, we could have gone back to iterating over
`m_profiles.values()` but `erase()` is slightly more efficient.

iamcodemaker committed:
remove unused deleteProfile() function

iamcodemaker committed:
save/load m_profileName when reading and writing

The code appears to assume that m_profileName is stored in the xml file,
but this wasn't actually the case leading to some issues. The profile
name was actually derrived from the filename. For profile names with `.`
characters in them specifically, anything after the first `.` was
stripped off.

iamcodemaker committed:
track the filename we used for broadcast profiles

Previously the filename was always derrived from the profile name which
causes problems if the names differ. For example, it was impossible to
remove a profile that had a `.` in the original profile name, after a
mixxx restart, unless you renamed the profile to the original name then
attempted the removal.

iamcodemaker commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -199,12 +191,17 @@ void BroadcastSettings::applyModel(BroadcastSettingsModel* pModel) {
     // Step 1: find profiles to delete from the settings
-    for(const auto& actualProfile : m_profiles) {
-        QString profileName = actualProfile->getProfileName();
+    for(auto actualProfile = m_profiles.begin(); actualProfile != m_profiles.end();) {
```
we could also use `m_profiles.values()` here, but using `erase()` is slightly more efficient.

uklotzde replied:
Even as a non-native speaker I suppose the term "actual" is wrong here. We should replace it. Since it is an iterator you may use "iterProfile" or whatever. I don't like type prefixes, but in this case it might be helpful to distinguish the iterator from the actual type, i.e. "removedProfile". Maybe "missingProfile" or "orphanedProfile" would be a more appropriate naming for this variable, because this is what we actually know about it before we are going to delete it.

Please add a space after the keyword _for_ to distinguish it from a function call.

iamcodemaker commented:
Thought of another edge case we should handle. If two names only differ by the special characters they have in them, they could resolve to the same file names and overwrite eachother.

iamcodemaker committed:
rename iterator

iamcodemaker committed:
fix whitespace after for

iamcodemaker committed:
don't attempt to delete the file if we never saved it

iamcodemaker committed:
reuse the old file name when saving a brodcast profile

If there was no old file name, generate a new one and make sure it
doesn't conflict with existing files.

iamcodemaker commented:
Updated to handle that edge case. I need to add some tests around this though.

iamcodemaker commented:
The `createProfile()` function is unused. Should I remove it as well?

iamcodemaker committed:
add tests for BroadcastSettings::saveProfile

iamcodemaker committed:
add test for BroadcastSettings::applyModel

iamcodemaker commented:
A thought I had, not for this PR, but maybe we should include default profiles for popular streaming services with Mixxx.

daschuer commented:
> A thought I had, not for this PR, but maybe we should include default profiles for popular streaming services with Mixxx.

This is a free advertising for them. So we should not do this unless there is money involved.  We can ask them to publish demo profiles on their domains though.
Do you have one in mind? Maybe it is worth to ask them.

iamcodemaker commented:
> This is a free advertising for them. So we should not do this unless there is money involved. We can ask them to publish demo profiles on their domains though.

Good point on the money. I had this thought as well. It would be better to charge money to be included as a default.

> Do you have one in mind? Maybe it is worth to ask them.

Yes, this came up in relation to https://cue.dj. Cue is a new (and beta) platform for streaming DJ sets. They post instructions for configuring Mixxx in their app and on their site.

daschuer commented:
Thank you for taking care here. 

This solution creates the situation that we store redundant info in file name and profile name. I am unsure but I think that it is possible now to have duplicate profiles names.
Can the filename change during the lifetime of the profile object? 

Can we find an unambiguous way to tun the profile name into a file name? Maybe a Git like hash code or something? Or revert storing the profile name and use a save encoding like the url encoding?
QUrl::toPercentEncoding()

What do you think?
